adapters/repos/db/queue/queue.go

```diff
@@ -0,0 +1,1063 @@
+	}, nil
+}
+
+func (q *DiskQueue) checkIfStale() (*chunk, error) {
+	if q.Size() == 0 {
+		return nil, nil
+	}
+
+	q.m.Lock()
+
+	if q.w.f == nil {
+		q.m.Unlock()
+		return nil, nil
+	}
+
+	if q.w.recordCount == 0 {
+		q.m.Unlock()
+		return nil, nil
+	}
+
+	if time.Since(q.lastPushTime) < q.staleTimeout {
+		q.m.Unlock()
+		return nil, nil
+	}
+
+	q.Logger.Debug("partial chunk is stale, scheduling")
+
+	err := q.w.Promote()
+	if err != nil {
+		q.m.Unlock()
+		return nil, err
+	}
+
+	q.m.Unlock()
+
+	return q.r.ReadChunk()
+}
+
+func (q *DiskQueue) Size() int64 {
+	if q == nil {
+		return 0
+	}
+
+	q.m.RLock()
+	defer q.m.RUnlock()
+
+	q.metrics.Size(q.recordCount)
+	q.metrics.DiskUsage(q.diskUsage)
+	return int64(q.recordCount)
+}
+
+func (q *DiskQueue) Pause() {
+	q.scheduler.PauseQueue(q.id)
+	q.metrics.Paused(q.id)
+}
+
+func (q *DiskQueue) Resume() {
+	q.scheduler.ResumeQueue(q.id)
+	q.metrics.Resumed(q.id)
+}
+
+func (q *DiskQueue) Wait() {
+	q.scheduler.Wait(q.id)
+}
+
+func (q *DiskQueue) Drop() error {
+	if q == nil {
+		return nil
+	}
+
+	err := q.Close()
+	if err != nil {
+		q.Logger.WithError(err).Error("failed to close queue")
+	}
+
+	q.m.Lock()
+	defer q.m.Unlock()
+
+	// remove the directory
+	err = os.RemoveAll(q.dir)
+	if err != nil {
+		return errors.Wrap(err, "failed to remove directory")
+	}
+
+	return nil
+}
+
+func (q *DiskQueue) removeChunk(c *chunk) {
+	q.rmLock.Lock()
+	defer q.rmLock.Unlock()
+
+	deleted, err := q.r.RemoveChunk(c)
+	if err != nil {
+		q.Logger.WithError(err).WithField("file", c.path).Error("failed to remove chunk")
+		return
+	}
+	if !deleted {
+		return
+	}
+
+	q.m.Lock()
+	q.recordCount -= c.count
+	q.diskUsage -= int64(c.size)
+	q.metrics.DiskUsage(q.diskUsage)
+	q.metrics.Size(q.recordCount)
+	q.m.Unlock()
+}
+
+// analyzeDisk is a slow method that determines the number of records
+// stored on disk and in the partial chunk by reading the header of all the files in the directory.
+// It also calculates the disk usage.
+// It is used when the queue is first initialized.
+func (q *DiskQueue) analyzeDisk() error {
+	q.m.Lock()
+	defer q.m.Unlock()
+
+	entries, err := os.ReadDir(q.dir)
+	if err != nil {
+		return errors.Wrap(err, "failed to read directory")
+	}
+
+	for _, entry := range entries {
+		if entry.IsDir() {
+			continue
+		}
+
+		// read the header of the chunk file
+		if chunkFilePattern.Match([]byte(entry.Name())) {
+			fi, err := entry.Info()
+			if err != nil {
+				return errors.Wrap(err, "failed to get file info")
+			}
+
+			if fi.Size() == 0 {
+				continue
+			}
+
+			q.diskUsage += fi.Size()
+
+			count, err := q.readChunkRecordCount(filepath.Join(q.dir, entry.Name()))
+			if err != nil {
+				return err
+			}
+
+			q.recordCount += count
+			continue
+		}
+	}
+
+	q.recordCount += q.w.recordCount
+
+	return nil
+}
+
+func (q *DiskQueue) readChunkRecordCount(path string) (uint64, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return 0, err
+	}
+	defer f.Close()
```

Comment:
shouldn't this `defer f.Close()` be put just after the 517 line so that the file handler would be also closed in case of error?

Reply:
In general it's better to do it after the err check because most of the time the APIs who return a ressource always return nil in case of error (and they are responsible for cleaning up)